### ISSUE TITLE
threads-required = "num-test-threads" may not be working on macos

### ISSUE THREAD

**Issue by jeromegn** [nextest-rs/nextest#831]
We have a CI building for both Linux and macOS. There's a test we need to run sequentially because it affects a global state.

The way we do this is with a `./config/nextest.toml` containing:

```toml
[profile.ci]
slow-timeout = "180s"
failure-output = "final"
fail-fast = false

[profile.ci.junit]
path = "junit.xml"

[[profile.ci.overrides]]
filter = "test(~wswg)"
threads-required = "num-test-threads"
```

Under Linux, it seems like it is doing the right thing, if I look at the junit output:

```xml
<testcase name="deploy_basic_node_app_w_wswg" classname="testflight::deploy_test" timestamp="2023-03-14T14:42:41.236+00:00" time="100.707"> </testcase>
<testcase name="machine_run" classname="testflight::deploy_test" timestamp="2023-03-14T14:44:21.947+00:00" time="5.915"> </testcase>
```

Under macOS:

```xml
<testcase name="deploy_basic_node_app_w_wswg" classname="testflight::deploy_test" timestamp="2023-03-14T14:42:41.039+00:00" time="110.462"> </testcase>
<testcase name="deploy_basic_node_app_w_remote_builder" classname="testflight::deploy_test" timestamp="2023-03-14T14:42:41.038+00:00" time="184.399"> 
```

On Linux, the tests are ran one after the other because the `~wswg` test uses all available test threads. I figured that out based on the timestamps and time to run the `~wswg` test.

On macOS, the tests are ran at the same time (again, based on the timestamps).

Did I configure it wrong?

**Comment by sunshowers** [comment]
Hi there, thanks for the report!

That's actually expected, currently: we currently have it so that the total number of threads can be exceeded in some situations. This isn't documented anywhere for nextest because we reserve the right to change this in the future, but it's mentioned in https://docs.rs/future-queue/latest/future_queue/#1-the-future_queue-adaptor.

If you want to ensure mutual exclusion for a subset of tests, you can use test groups: https://nexte.st/book/test-groups.html. However I'm not sure exactly what you want here. Do you want this test to be run separately from *all* other tests in your workspace? If that's the case, we'd have to change future-queue so that it can never exceed the overall weight. (Which is fine, I've been thinking about doing this anyway.)

**Comment by jeromegn** [comment]
Interesting. I tried test groups first but it didn't do what I wanted (running this test separately from all others).

We have a setup on buildkite that uses parallelism, I'm not sure if that can interfere too.

**Comment by sunshowers** [comment]
I'm going to try and get a release out this weekend that enables your use case.

**Comment by sunshowers** [comment]
Just released support for this in cargo-nextest 0.9.51. Your use case has also been mentioned on the page for [`threads-required`](https://nexte.st/book/threads-required.html#use-cases).

**Comment by jeromegn** [comment]
Thank you! That does fix it for us.
